8324641: [IR Framework] Add Setup method to provide custom arguments and set fields (#17557)
openjdk/jdk: test/hotspot/jtreg/compiler/lib/ir_framework/README.md

## 2. Features
The framework offers various annotations and flags to control how your test code should be invoked and being checked. This section gives an overview over all these features.

### 2.1 Different Tests

> There are two ways a test can be formulated, depending on how much control is needed over the test invocation.

Reviewer: ```suggestion
There are two ways a test can be written, depending on how much control is needed over the test invocation.
```

The framework offers various annotations and flags to control how your test code should be invoked and being checked. This section gives an overview over all these features.

### 2.1 Different Tests
There are two ways a test can be written, depending on how much control is needed over the test invocation.

#### Normal Test
The normal and simplest form of testing provides a single `@Test` annotated method which the framework invokes directly as part of the testing. The test method either has no arguments, or they must be specified with an `@Arguments` annotation.

Arguments can be provided with `@Arguments(values = {...})` by providing well-specified inputs for each individual argument. Alternatively, a setup method can be chosen with `@Arguments(setup = "setupMethodName")`, which computes arguments and can also set fields.

> More information on normal test methods with a precise definition can be found in the Javadocs of [Test](./Test.java). Concrete examples on how to specify a normal test can be found in [BaseTestsExample](../../../testlibrary_tests/ir_framework/examples/BaseTestExample.java).

Reviewer: Since we now change the README and the definition of two kinds of tests, I also suggest to rename `BaseTestExample.java` to `NormalTestExample.java`.

Author: done

##### Setup Method

> A `@Setup` annotated method can provide custom arguments and set fields before a normal test is run. A normal test method can be annotated with `@Arguments(setup = "setupMethodName")` to specify which setup method is to be used.

Reviewer: ```suggestion
A `@Setup` annotated method can provide custom arguments and set fields before a normal test is run. A `@Test` annotated method can additionally be annotated with `@Arguments(setup = "setupMethodName")` to define the dedicated `@Setup` method.
```

> More information on checked tests with a precise definition can be found in the Javadocs of [Setup](./Setup.java). Concrete examples on how to specify a setup method can be found in [SetupExample](../../../testlibrary_tests/ir_framework/examples/SetupExample.java).

Reviewer: ```suggestion
More information on normal tests with `@Setup` methods together with a precise definition can be found in the Javadocs of [Setup](./Setup.java). Concrete examples on how to specify a setup method can be found in [SetupExample](../../../testlibrary_tests/ir_framework/examples/SetupExample.java).
```

##### Check Method
A `@Check` annotated method is invoked directly after a corresponding normal `@Test` annotated method. The user can perform various checks, such as test method return value and field value verification.

> More information on check methods with a precise definition can be found in the Javadocs of [Check](./Check.java). Concrete examples on how to specify check methods can be found in [CheckedTestsExample](../../../testlibrary_tests/ir_framework/examples/CheckedTestExample.java).

Reviewer: Maybe add here an additional paragraph:

`@Setup` and `@Check` methods can only be specified for normal but not for custom run tests (see next section).

Author: done

#### Custom Run Tests

> The normal test method is invoked directly by the framework, and hence the user does not have control over the invokation.

Author: It was to mirror the old description of Custom Run Test:
"Neither the base nor the checked tests provide any control over..."
"A custom run test gives full control..."

> The normal test method is invoked directly by the framework, and hence the user does not have control over the invokation.

Author: But I can remove it if you think that it is not really helpful to make the contrast.

> A custom run test gives full control over the invocation of the `@Test` annotated method to the user. The framework calls a dedicated `@Run` annotated method from which the user can invoke the `@Test` method according to their needs.

Reviewer: I suggest to add some more details what can be done and should be done (i.e. setup and verification):

```suggestion
A custom run test gives full control over the invocation of the `@Test` annotated method to the user which includes argument and field setup as well as result and field value verification. The framework calls a dedicated `@Run` annotated method from which the user can invoke the `@Test` method according to their needs.
```
